aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,218 @@
+            message='The parent folder did not contain the required XML output file.')
+        spec.exit_code(310, 'ERROR_OUTPUT_STDOUT_READ',
+            message='The stdout output file could not be read.')
+        spec.exit_code(312, 'ERROR_OUTPUT_STDOUT_INCOMPLETE',
+            message='The stdout output file was incomplete.')
+        spec.exit_code(340, 'ERROR_OUT_OF_WALLTIME_INTERRUPTED',
+            message='The calculation stopped prematurely because it ran out of walltime but the job was killed by the '
+                    'scheduler before the files were safely written to disk for a potential restart.')
+        spec.exit_code(350, 'ERROR_UNEXPECTED_PARSER_EXCEPTION',
+            message='The parser raised an unexpected exception.')
+
+        # Output datafile related exceptions
+        spec.exit_code(330, 'ERROR_OUTPUT_DATAFILE_MISSING',
+            message='The retrieved folder did not contain the required formatted data output file.')
+        spec.exit_code(331, 'ERROR_OUTPUT_DATAFILE_READ',
+            message='The formatted data output file could not be read.')
+        spec.exit_code(332, 'ERROR_UNSUPPORTED_DATAFILE_FORMAT',
+            message='The data file format is not supported by the parser')
+
+    def prepare_for_submission(self, folder): # pylint: disable=too-many-branches,too-many-statements
+        """Create the input files from the input nodes passed to this instance of the `CalcJob`.
+
+        :param folder: an `aiida.common.folders.Folder` to temporarily write files on disk
+        :return: `aiida.common.datastructures.CalcInfo` instance
+        """
+
+        # Put the first-level keys as uppercase (i.e., namelist and card names) and the second-level keys as lowercase
+        if 'parameters' in self.inputs:
+            parameters = _uppercase_dict(self.inputs.parameters.get_dict(), dict_name='parameters')
+            parameters = {k: _lowercase_dict(v, dict_name=k) for k, v in six.iteritems(parameters)}
+        else:
+            parameters = {}
+
+        # Set default values. NOTE: this is different from PW/CP
+        for blocked in self._blocked_keywords:
+            namelist = blocked[0].upper()
+            key = blocked[1].lower()
+            value = blocked[2]
+
+            if namelist in parameters:
+                if key in parameters[namelist]:
+                    raise exceptions.InputValidationError(
+                        "You cannot specify explicitly the '{}' key in the '{}' "
+                        'namelist.'.format(key, namelist))
+            else:
+                parameters[namelist] = {}
+            parameters[namelist][key] = value
+
+        # Check for essential keys
+        try:
+            plot_num = parameters['INPUTPP']['plot_num']
+            dimension = parameters['PLOT']['iflag']
+        except KeyError:
+            raise exceptions.InputValidationError("Both 'plot_num' and 'iflag' must be set")
```

Comment:
since they are in different namelists, would it maybe be better to add those explicitly in the exception message?

Reply:
I separated them into different try/excepts just.